source/RemoteWiring/HardwareProfile.cpp

```diff
@@ -0,0 +1,381 @@
+        break;
+    }
+}
+
+HardwareProfile::HardwareProfile(
+    int total_number_of_pins_,
+    int number_of_analog_pins_
+    ) :
+    _is_valid( ATOMIC_VAR_INIT( false ) ),
+    _total_pins( ATOMIC_VAR_INIT( total_number_of_pins_ ) ),
+    _analog_offset( ATOMIC_VAR_INIT( total_number_of_pins_ - number_of_analog_pins_ ) ),
+    _num_analog_pins( ATOMIC_VAR_INIT( number_of_analog_pins_ ) ),
+    _pinCapabilities( nullptr ),
+    _analogResolutions( nullptr ),
+    _pwmResolutions( nullptr ),
+    _servoResolutions( nullptr )
+{
+}
+
+
+HardwareProfile::~HardwareProfile()
+{
+    _is_valid = false;
+
+    if( _pinCapabilities != nullptr )
+    {
+        delete( _pinCapabilities );
+    }
+
+    if( _analogResolutions != nullptr )
+    {
+        delete( _analogResolutions );
+    }
+
+    if( _pwmResolutions != nullptr )
+    {
+        delete( _pwmResolutions );
+    }
+
+    if( _servoResolutions != nullptr )
+    {
+        delete( _servoResolutions );
+    }
+}
+
+
+//******************************************************************************
+//* Public Methods
+//******************************************************************************
+
+uint8_t
+HardwareProfile::getPinCapabilitiesBitmask(
+    uint8_t pin_
+    )
+{
+    if( !_is_valid || _pinCapabilities == nullptr || pin_ < 0 || pin_ >= _total_pins )
+    {
+        return 0;
+    }
+    return _pinCapabilities->at( pin_ );
+}
+
+bool
+HardwareProfile::isAnalogSupported(
+    uint8_t pin_
+    )
+{
+    return ( getPinCapabilitiesBitmask( pin_ ) & static_cast<uint8_t>( PinCapability::ANALOG ) ) > 0;
+}
+
+bool
+HardwareProfile::isDigitalInputSupported(
+    uint8_t pin_
+    )
+{
+    return ( getPinCapabilitiesBitmask( pin_ ) & static_cast<uint8_t>( PinCapability::INPUT ) ) > 0;
+}
+
+bool
+HardwareProfile::isDigitalInputPullupSupported(
+    uint8_t pin_
+    )
+{
+    return ( getPinCapabilitiesBitmask( pin_ ) & static_cast<uint8_t>( PinCapability::INPUT_PULLUP ) ) > 0;
+}
+
+bool
+HardwareProfile::isDigitalOutputSupported(
+    uint8_t pin_
+    )
+{
+    return ( getPinCapabilitiesBitmask( pin_ ) & static_cast<uint8_t>( PinCapability::OUTPUT ) ) > 0;
+}
+
+bool
+HardwareProfile::isI2cSupported(
+    uint8_t pin_
+    )
+{
+    return ( getPinCapabilitiesBitmask( pin_ ) & static_cast<uint8_t>( PinCapability::I2C ) ) > 0;
+}
+
+bool
+HardwareProfile::isPwmSupported(
+    uint8_t pin_
+    )
+{
+    return ( getPinCapabilitiesBitmask( pin_ ) & static_cast<uint8_t>( PinCapability::PWM ) ) > 0;
+}
+
+bool
+HardwareProfile::isServoSupported(
+    uint8_t pin_
+    )
+{
+    return ( getPinCapabilitiesBitmask( pin_ ) & static_cast<uint8_t>( PinCapability::SERVO ) ) > 0;
+}
+
+//******************************************************************************
+//* Private Methods
+//******************************************************************************
+
+void
+HardwareProfile::initializeWithFirmata(
+    Windows::Storage::Streams::IBuffer ^buffer_
```

Comment:
this was done for consistency with the rest of the code. We can discuss changing this, but we'll want to experiment first with how it changes the API.